ddtrace/internal/forksafe.py

```diff
@@ -55,3 +57,55 @@ def _after_fork():
         ddtrace_after_in_child()
 
     threading._after_fork = _after_fork  # type: ignore[attr-defined]
+
+
+_resetable_objects = weakref.WeakSet()  # type: weakref.WeakSet[ResetObject]
+
+
+def _reset_objects():
+    # type: (...) -> None
+    for obj in _resetable_objects:
+        try:
+            obj._reset_object()
+        except Exception:
+            log.exception("Exception ignored in object reset forksafe hook %r", obj)
+
+
+register(_reset_objects)
+
+
+class ResetObject(wrapt.ObjectProxy):
+    """An object wrapper object that is fork-safe and resets itself after a fork.
+
+    When a Python process forks, a Lock can be in any state, locked or not, by any thread. Since after fork all threads
+    are gone, Lock objects needs to be reset. CPython does this with an internal `threading._after_fork` function. We
+    use the same mechanism here.
+
+    """
+
+    def __init__(
+        self, wrapped_class  # type: typing.Any
+    ):
+        # type: (...) -> None
+        super(ResetObject, self).__init__(wrapped_class())
+        self._self_wrapped_class = wrapped_class
+        _resetable_objects.add(self)
+
+    def _reset_object(self):
+        # type: (...) -> None
```

Comment:
I wonder if we should have some kind of interface such that objects can control their recreation if needed (ex: `TracerWriter` defines a `recreate` API which is to essentially handle reinitializing after a fork). Might be handy if some attributes need to be persisted.

Definitely not required for now and we can add it if/when necessary but wanted to brain dump here if there are any future readers.

Reply:
https://github.com/DataDog/dd-trace-py/pull/2245 will help for this though is there something other than the threading that you have in mind here @Kyle-Verhoog ?

Reply:
Oh yes, looks like that would do it